# 2) Find the duplicates!

#### Hash Functions and Hash Tables
#### Our Algorithm for non-ordered case:
First, We define a create_alphabet_table(alphabet) function that get as an input the alphapet and pecial characters as a list, for each character, it give a number and then as an output it will gives an alphabet_table, we calculated each character number by using the hash function in python and produce it by a prime number that we store in primes1 list. Formula:
$$ alphabettable[c]=abs(hash(c) · prime)$$
Then, we define our first hash function,string to number, that convert the string to a number by summing the numbers of string's characters from the alphabet_table.
$$paswordhash= \sum_{c=1}^n alphabettable[c]$$
Finally, we define a second hash function that apply this algorithm: number_to_hash_I function
##### . Choose a prime p betwen M and 2*M, where M is the interval of our strings' numbers "keys"
##### . Then the hash function will be:
$$h(k) = ((a·k+b) mod p) mod N$$
where k is the number of the string, a is a random number: 0<a<p and b a random number: 0≤b<p
and N is the length of the hash table, we chose the prime number 99999999977.
Then for each password we convert it to a number then we hash it and store it as a sorted string in a text file with two columns: password, hash, as you can see in the function: 
#### Our Algorithm for ordered case:
tHis algorithm different from previous case just in string to number function where here we will sum the numbers of string's characters from the alphabet_table * index of the character in the string, considering that the string index beginning from 2 just to increase the effect of the character order,$$paswordhash= \sum_{c=1}^n alphabettable[c]· c$$ and also we don't sort the string's characters before storing as in the other case.
#### False Positive:
To find the false positive, we define "FB" function that read the text files that we got in hashing part, then build a dictionary "FB_d" that contain the hash values as keys and a set of passwords that have same hash as values, then after building this dictionary we add the length - 1 of each set contain more than two values to "FB_Count" variable, considering that each set contain len(set)-1 collisions with the first value of the set.

In [3]:
import math
import numpy as np
import random

In [4]:
from random import shuffle

In [53]:
# This function to check if the number is prime or not
def is_prime(number):
    for j in range(2, int(math.sqrt(number)+1)):
        if (number % j) == 0:
            return False
    return True

In [ ]:
#creating a file with prime numbers between M and 2*M, P
with open("outprimesfile.txt", "w") as text_file:
    for i in range(700000000000000000000,1400000000000000000000):
        if is_prime(i):
            text_file.write(str(i)+'\n')
text_file.close()

In [6]:
#This file "outprimesff" contain a 9 digits prime numbers which will we use to create the alphabet hash list
primes1=[]
with open("outprimesff.txt", "r") as r:
        for prime in r:
            primes1.append(prime.replace('\n',''))
r.close()
shuffle(primes1)

In [7]:
# primes list which contain possible values of P
primes=[]
with open("outprimesfile.txt", "r") as r:
        for prime in r:
            primes.append(prime.replace('\n',''))
r.close()
shuffle(primes)

In [8]:
#Defining the alphabet letters and special characters
Alphabet=['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z','a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z','1','2','3','4','5','6','7','8','9','0','<','>','\\','/','!','@','#','$','%','^','*','&','.',',','"','\'',':',';','(',')','?','+','=','-']

In [9]:
#give each letter a number to use in converting string to large number process
def create_alphabet_table(alphabet):
    alpha=dict()
    for c in alphabet:
        alpha[c]=abs(hash(c)*int(primes1[random.randint(0, len(primes1)-2)]))
    return alpha
#converting password string to a large number, non order case
def string_to_number_ignoring_order(password,alphabet_table):
    pasword_hash=0
    for c in password:
        pasword_hash+=alphabet_table[c]
    return pasword_hash

In [11]:
alphabet_table=create_alphabet_table(Alphabet)

In [14]:
pd.DataFrame.from_dict(alphabet_table,orient='index').head()

0
A  12869310408512279817
B  35678008092121553746
C  22503307280597754940
D  40362414180771602880
E  43601387269565635688

In [59]:
string_to_number_ignoring_order('aaba',alphabet_table)

134972597154273191913

In [60]:
string_to_number_ignoring_order('aaab',alphabet_table)

134972597154273191913

In [61]:
#number to hash function, non order case, here we sort the password before storing
def number_to_hash_I():
    with open("passwords.txt", "r") as r,open("passwordshashnumbersIII.txt", "w") as rr:
        for password in r:
            string_to_number=string_to_number_ignoring_order(password.replace('\n',''),alphabet_table)
            p =int(primes[random.randint(0, len(primes)-1)])
            a = random.randint(1, p-1)
            b = random.randint(0, p-1)
            number_to_hash=((a*string_to_number+b) % p) % 99999999977
            password1=str(''.join(sorted(password.replace('\n',' '))))
            rr.write(password1+' '+str(number_to_hash)+'\n')

In [ ]:
number_to_hash_I()

In [62]:
#converting password string to a large number, order case
def string_to_number_considering_order(password,alphabet_table):
    pasword_hash=0
    j=2
    for c in password:
        pasword_hash+=alphabet_table[c]*j
        j+=1
    return pasword_hash

In [65]:
string_to_number_considering_order('aaba',alphabet_table)

461562574396820114828

In [66]:
string_to_number_considering_order('aaab',alphabet_table)

439879543110548001093

In [68]:
#number to hash, consider order case, here we don't sorted the password before storing
def number_to_hash_C():
    with open("passwords.txt", "r") as r,open("passwordshashnumbersCCC.txt", "w") as rr:
        for password in r:
            string_to_number=string_to_number_considering_order(password.replace('\n',''),alphabet_table)
            p =int(primes[random.randint(0, len(primes)-1)])
            a = random.randint(1, p-1)
            b = random.randint(0, p-1)
            number_to_hash=((a*string_to_number+b) % p) % 99999999977
            rr.write(password.replace('\n','')+' '+str(number_to_hash)+'\n')

In [ ]:
number_to_hash_C()

In [1]:
import pandas as pd
import csv

In [2]:
#this fnction is to read the txt files that we got as output from number_to_hash_I() and number_to_hash_C() functions
def data_aggregator(path,chunksize,columnname):
    result = pd.DataFrame()
    for chunk in pd.read_csv(path,sep=' ',chunksize=chunksize,names=['password','hash'],encoding='utf-8',quoting=csv.QUOTE_NONE):
        result=result.append(pd.DataFrame(chunk[columnname]))
    return result

### 1. non ordered case

In [31]:
#reading the password and hash columns from the txt file that produced from number_to_hash_I() function
df=data_aggregator("passwordshashnumbersIII.txt",10000,['password','hash'])

In [34]:
df.head()

password         hash
0  "$-/3DHIOQTUYcdefqrv  26951500055
1  "$&(*+.36AMOQUaeortx  35990329136
2  '*-/6;HMNPThjkloqruw  20102137125
3  "%*3579@ACJMSbefhlux  58195741152
4  "#'5;=BCIJLSbijmpuxy   2013483790

In [69]:
df=None
del df

#### Password

In [25]:
#reading the password column from the txt file that produced from number_to_hash_I() function
df=data_aggregator("passwordshashnumbersIII.txt",10000,'password')

##### Find the duplicates count of password column

In [26]:
d=df.groupby(df.password.tolist(),as_index=False).size().reset_index(name='Count')

In [27]:
dup_count_i=d.loc[d.Count>1].sum()

In [29]:
password_dup_i=dup_count_i.Count

In [30]:
password_dup_i

10000000

In [70]:
df=None
del df
d=None
del d

### FB count

In [14]:
'''this function is to count the FB'''
'''we directly count the fb we don't count the real duplicates'''
def FB(filename):
    FB_Count=0
    FB_d=dict()
    with open(filename, "r") as r:
        for line in r:
            hashh=line.split(' ')[1].replace('\n','')
            password=line.split(' ')[0].replace('\n','')
            if hashh not in FB_d:
                FB_d[hashh]=[password]
            elif password not in FB_d[hashh]:
                FB_d[hashh].append(password)
    for key in FB_d.keys():
        if len(FB_d[key])>1:
            FB_Count+=len(FB_d[key])-1
    return FB_Count,FB_d

In [48]:
FB('passwordshashnumbersIII.txt')

3028

In non ordered case: duplicates=10000000,FB=3028

### 2. ordered case

In [37]:
#reading the password and hash columns from the txt file that produced from number_to_hash_C() function
df=data_aggregator("passwordshashnumbersCCC.txt",10000,['password','hash'])

In [39]:
df.head()

password         hash
0  OHcv-/U3QI$rdqYTef"D  32385060987
1  QtA*.xM$e(+"aO36r&Uo  58482258375
2  T;rqw/ou'HN-Pklj6hM*  80960637150
3  b%xJ79"A*C5@ehMfS3lu   3957624150
4  buI=;LpjBiCm"JS5'#xy  79995253771

#### Passwords

In [40]:
#reading the password column from the txt file that produced from number_to_hash_C() function
dfc=data_aggregator("passwordshashnumbersCCC.txt",10000,'password')

##### Find the duplicates count of password column

In [41]:
dc=dfc.groupby(dfc.password.tolist(),as_index=False).size().reset_index(name='Count')

In [42]:
dup_count_c=dc.loc[dc.Count>1].sum()

In [43]:
password_dup_c=dup_count_c.Count

In [45]:
password_dup_c

5000000

In [46]:
dc=None
del dc
dfc=None
del dfc

### FB count

In [47]:
FB('passwordshashnumbersCCC.txt')

0

In ordered case: duplicates=5000000,FB=0

#### -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------